cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/DefaultKafkaQuotasManager.java

```diff
@@ -0,0 +1,226 @@
+     * This is used in case when we need to reset the default Kafka quotas
+     *
+     * @return  empty instance of the {@link QuotasPluginKafka}
+     */
+    private static QuotasPluginKafka emptyQuotasPluginKafka() {
+        QuotasPluginKafka quotasPluginKafka = new QuotasPluginKafka();
+
+        quotasPluginKafka.setConsumerByteRate(null);
+        quotasPluginKafka.setControllerMutationRate(null);
+        quotasPluginKafka.setProducerByteRate(null);
+        quotasPluginKafka.setRequestPercentage(null);
+
+        return quotasPluginKafka;
+    }
+
+    /**
+     * Creates list of {@link ClientQuotaAlteration.Op} based on the configuration in {@link QuotasPluginKafka}.
+     * This list will be then passed to the Admin client and the default user quotas will be set.
+     *
+     * @param quotasPluginKafka     configuration of the quotas
+     *
+     * @return  list of {@link ClientQuotaAlteration.Op}
+     */
+    private static List<ClientQuotaAlteration.Op> prepareQuotaConfigurationRequest(QuotasPluginKafka quotasPluginKafka) {
+        List<ClientQuotaAlteration.Op> ops = new ArrayList<>();
+
+        ops.add(new ClientQuotaAlteration.Op(PRODUCER_BYTE_RATE_QUOTA,
+            quotasPluginKafka.getProducerByteRate() != null ? Double.valueOf(quotasPluginKafka.getProducerByteRate()) : null));
+        ops.add(new ClientQuotaAlteration.Op(CONSUMER_BYTE_RATE_QUOTA,
+            quotasPluginKafka.getConsumerByteRate() != null ? Double.valueOf(quotasPluginKafka.getConsumerByteRate()) : null));
+        ops.add(new ClientQuotaAlteration.Op(REQUEST_PERCENTAGE_QUOTA,
+            quotasPluginKafka.getRequestPercentage() != null ? Double.valueOf(quotasPluginKafka.getRequestPercentage()) : null));
+        ops.add(new ClientQuotaAlteration.Op(CONTROLLER_MUTATION_RATE_QUOTA,
+            quotasPluginKafka.getControllerMutationRate() != null ? quotasPluginKafka.getControllerMutationRate() : null));
+
+        return ops;
+    }
+
+    /**
+     * Based on configuration in {@param quotasPlugin}, it configures the default user quota in Kafka.
+     *
+     * @param reconciliation            Reconciliation marker
+     * @param vertx                     Vert.x instance
+     * @param adminClientProvider       Kafka Admin client provider
+     * @param pemTrustSet               Trust set for TLS authentication in PEM format
+     * @param pemAuthIdentity           Identity for TLS client authentication in PEM format
+     * @param quotasPlugin              Configuration of Kafka quotas plugin
+     *
+     * @return  Future that completes when the default user quota configuration is completed
+     */
+    public static Future<Void> reconcileDefaultUserQuotas(
+        Reconciliation reconciliation,
+        Vertx vertx,
+        AdminClientProvider adminClientProvider,
+        PemTrustSet pemTrustSet,
+        PemAuthIdentity pemAuthIdentity,
+        QuotasPlugin quotasPlugin
+    ) {
+        Promise<Void> promise = Promise.promise();
```

Comment:
Should this promise be created in the scope where it is used? Or is it even needed? It seems you can simple return a future directly in the places where you use it.

Reply:
I think that the promise is useful for the `onComplete` method there (because I cannot simply return the Future from there, but with the promise I can handle the exceptions etc.), but I agree that I should move it under the scope where it is needed and handle the situations there correctly. Thanks

Reply:
You can replace it with `.compose`.

Reply:
In this situation, I just want to handle the result of the previous call (if it was successful or not) -> how it is done on other places. Or does the `.compose` makes more sense here?

Reply:
I think it will help to properly understand what is done when and where. Because you will get additional checks for returning the right values etc.

Reply:
Okay, I thought that when the `alterQuotas` returns `Future<Void>` there is not much to check -> or I can maybe use `.recover` alternatively. But yeah I can do some check for the result value or maybe return `Future<Boolean>` there and handle the checks accordingly.

Reply:
You can either return the result of the alter quotas. Or do something like this if you want the log messages:

```
                    return alterQuotas(reconciliation, vertx, kafkaAdmin, clientQuotaAlteration)
                        .compose(result -> {
                            if (result.succeeded()) {
                                LOGGER.infoCr(reconciliation, "Successfully altered default user quotas");
                                kafkaAdmin.close();
                                return Future.succeededFuture();
                            } else {
                                LOGGER.warnCr(reconciliation, "Failed to alter default user quotas", result.cause());
                                kafkaAdmin.close();
                                return Future.failedFuture(result.cause());
                            }
```

I think that will make it easier for anyone reading the code to reason what and when is done and what are all the places there the chain ends. Its not like your code would not work, I just wonder if this is more readable.

If you wanna keep the Promise and the onComplete, fine. But the original part of the comment still stands, the promise should be created only in the scope where it is needed / used.

Reply:
> Or do something like this if you want the log messages

Yeah I wanted to do that this way, however the `result` here doesn't have the `succeeded()` and `failed()` methods, I can just check that the result equals to something (in this case `null`).
(it's because the `result` is `Void` type and it's not a handler)

So I will maybe proceed with the promise and have it correctly handled (and right scope). Thanks

Reply:
Right ... that is what happens when you type code out of memory and not in an IDE :-D ... so you can still use recover or how is it called. But as I said, if you wanna keep the onComplete, fine. But fix the scope.

Reply:
> Right ... that is what happens when you type code out of memory and not in an IDE :-D .

No problem 😄 it's great that you are able to remember everything actually 😄 

> But fix the scope.

Will do :)